src/Primitives.hs

```diff
@@ -221,53 +217,63 @@ primitiveImplements x@(XObj _ _ _) ctx args =
-              Just (XObj (Lst interfaces) _ _) ->
-                do
-                  let result = foldM (\ctx' (xobj, interface) -> registerInInterface ctx' xobj interface) ctx (zip (cycle [annXObj]) (map getPath interfaces))
-                  case result of
-                    Left err ->
-                      do
-                        case contextExecMode ctx of
-                          Check ->
-                            let fppl = projectFilePathPrintLength (contextProj ctx)
-                             in putStrLn (machineReadableInfoFromXObj fppl annXObj ++ " " ++ err)
-                          _ -> putStrLnWithColor Red err
-                        pure ctx
-                    Right ctx' -> pure (ctx' {contextGlobalEnv = envInsertAt globalEnv (getPath annXObj) (Binder adjustedMeta annXObj)})
-              _ -> pure (ctx {contextGlobalEnv = envInsertAt globalEnv (getPath annXObj) (Binder adjustedMeta annXObj)})
+  pure (hideIt freshBinder)
+    >>= \newBinder -> case annXObj of
+      XObj (Lst (XObj (Defalias _) _ _ : _)) _ _ ->
+        pure (ctx {contextTypeEnv = TypeEnv (envInsertAt (getTypeEnv typeEnv) (getPath annXObj) newBinder)})
+      XObj (Lst (XObj (Deftype _) _ _ : _)) _ _ ->
+        pure (ctx {contextTypeEnv = TypeEnv (envInsertAt (getTypeEnv typeEnv) (getPath annXObj) newBinder)})
+      XObj (Lst (XObj (DefSumtype _) _ _ : _)) _ _ ->
+        pure (ctx {contextTypeEnv = TypeEnv (envInsertAt (getTypeEnv typeEnv) (getPath annXObj) newBinder)})
+      _ -> defineGlobal
+  where
+    freshBinder = (Binder emptyMeta annXObj)
+    defineGlobal :: IO Context
+    defineGlobal =
+      when (projectEchoC proj) (putStrLn (toC All (Binder emptyMeta annXObj)))
+        >> pure (lookupInEnv (getPath annXObj) globalEnv)
+        >>= \maybebinder ->
+          case maybebinder of
+            Nothing -> pure (hideIt freshBinder)
+              >>= \newBinder -> pure (ctx {contextGlobalEnv = envInsertAt globalEnv (getPath annXObj) newBinder})
+            Just (_, (Binder meta _)) -> pure (hideIt (Binder meta annXObj))
+              >>= \newBinder -> warnTypeChange newBinder
+                >> registerDef newBinder
+                >>= \context -> pure (context {contextGlobalEnv = envInsertAt globalEnv (getPath annXObj) newBinder})
+    hideIt :: Binder -> Binder
+    hideIt binder =
+      if hidden
+        then Meta.updateBinderMeta binder "hidden" trueXObj
+        else binder
+    warnTypeChange :: Binder -> IO ()
+    warnTypeChange binder =
+      unless (areUnifiable (forceTy annXObj) previousType) warn
+      where
+        previousType = forceTy (binderXObj binder)
+        warn =
+          emitWarning
+            ( "Definition at " ++ prettyInfoFromXObj annXObj ++ " changed type of '" ++ show (getPath annXObj)
+                ++ "' from "
+                ++ show previousType
+                ++ " to "
+                ++ show (forceTy annXObj)
+            )
+    registerDef :: Binder -> IO Context
```

Comment:
The word "register" makes me think of the `register` for external functions/types, so maybe we can use something less misleading here?

Reply:
Addendum: Noticed it being used above too. My loose convention has been to use "register" for external stuff and "define" for top level forms. So maybe these should be called something with define too.

Reply:
sounds good--it refers to adding the xobj to an interface, which historically was also called registerInInterface...--let's rename it `implementInterface`?